docs/src/main/asciidoc/security-openid-connect-client.adoc

```diff
@@ -0,0 +1,218 @@
+
+You can selectively register `OidcClientRequestFilter` by using either `io.quarkus.oidc.rest.client.OidcClientFilter` or `org.eclipse.microprofile.rest.client.annotation.RegisterProvider` annotations:
+
+```
+import org.eclipse.microprofile.rest.client.inject.RegisterRestClient;
+import io.quarkus.oidc.rest.client.OidcClientFilter;
+
+@RegisterRestClient
+@OidcClientFilter
+@Path("/")
+public interface ProtectedResourceService {
+
+    @GET
+    String getUserName();
+}
+```
+
+or
+
+```
+import org.eclipse.microprofile.rest.client.annotation.RegisterProvider;
+import org.eclipse.microprofile.rest.client.inject.RegisterRestClient;
+import io.quarkus.oidc.rest.client.OidcClientRequestFilter;
+
+@RegisterRestClient
+@RegisterProvider(OidcClientRequestFilter.class)
+@Path("/")
+public interface ProtectedResourceService {
+
+    @GET
+    String getUserName();
+}
+```
+
+Alternatively, `OidcClientRequestFilter` can be registered automatically with all MP Rest or JAX-RS clients if `quarkus.oidc.rest.client.register-filter=true` property is set. 
+
+
+== Use injected GrantTokens
+
+If you prefer you can use your own custom filter and inject `GrantTokens`:
+
+```
+@Provider
+@Priority(Priorities.AUTHENTICATION)
+@RequestScoped
+public class OidcClientRequestCustomFilter implements ClientRequestFilter {
+
+    @Inject
+    GrantTokens grantTokens;
+
+    @Override
+    public void filter(ClientRequestContext requestContext) throws IOException {
+        requestContext.getHeaders().add(HttpHeaders.AUTHORIZATION, "Bearer " + grantTokens.getAccessToken());
+    }
+}
+```
+
+The `GrantTokens` producer will acquire and refresh the tokens and the custom filter will decide how and when to use the token.
+
+See also the previous section about delaying the token acquisition in some cases.
+
+== OidcClients
+
+`io.quarkus.oidc.client.OidcClient` is a container of `OidcClient`s - it includes a default `OidcClient` (which can also be injected directrly as described above) and named clients which can be configured like this:
+
+```
+quarkus.oidc-client.client-enabled=false
+
+quarkus.oidc-client.jwt-secret.auth-server-url=${keycloak.url}/realms/quarkus2/
+quarkus.oidc-client.jwt-secret.client-id=quarkus-app
+quarkus.oidc-client.jwt-secret.credentials.jwt.secret=AyM1SysPpbyDfgZld3umj1qzKObwVMkoqQ-EstJQLr_T-1qS0gZH75aKtMN3Yj0iPS4hcgUuTwjAzZr1Z9CAow
+```
+
+Note in this case the default client is disabled with a `client-enabled=false` property. The `jwt-secret` client can be accessed like this:
+
+[source,java]
+----
+import javax.inject.Inject;
+import javax.ws.rs.GET;
+import javax.ws.rs.Path;
+
+import io.quarkus.oidc.client.OidcClient;
+import io.quarkus.oidc.client.OidcClients;
+
+@Path("/clients")
+public class OidcClientResource {
+
+    @Inject
+    OidcClients clients;
+
+    @GET
+    public String getResponse() {
+        OidcClient client = clients.getClient("jwt-secret");
```

Comment:
Maybe this would be nicer if you provide a `@OidcClient` annotation where you can define a name and have the named  client injected.

Reply:
Nice idea, I'll have a look; I'll keep `OidcClients` as I've already prototyped creating the clients on demand by copying/pasting your code from quarkus-oidc with a few cosmetic updates :-)

Reply:
@pedroigor Hi Pedro, Martin @mkouba explained it is not going to work for ApplicationScoped endpoints, so for now I'll have to resolve this comment without any action. Using `OidcClients` is a bit more verbose to get a named client but I suppose it is a more complete approach - as it also allows to create the clients dynamically; As I suggested in the docs, if the users have multiple named clients then, at least when they have OIDC multi-tenancy, then can use a tenant id for a client look up, this selection of the named clients is supposed to be dynamic I guess; thanks, nice idea though :-)